### ISSUE TITLE
McLennan: set creep speed = JVEL in controller

### ISSUE THREAD

**Issue by Tom-Willemsen** [ISISComputingGroup/IBEX#4815]
As a user of the McLennan motor controller, I want internal homes (mode 2) to run at the same speed as external homes.

When in internal homing mode, the McLennan uses the "creep speed" to determine the speed of this home. IBEX currently does not set this speed at all, meaning it could potentially be extremely slow.

The relevant command is, for example:
```
1SC100\r\n
```
Meaning axis 1, "set creep speed", 100.

The McLennan driver should set this creep speed to be equal to the jog speed (JVEL).

Note: the speed was manually sent to the controller on MERLIN to solve https://github.com/ISISComputingGroup/IBEX/issues/4740 but will possibly be forgotten on controller power-cycle. This is the long-term solution to that ticket.

**Comment by DominicOram** [comment]
Once complete remove documentation for temporary fix https://github.com/ISISComputingGroup/ibex_developers_manual/wiki/McLennan-motors#homes-are-very-slow

**Comment by Tom-Willemsen** [comment]
There is also a command to get it to save settings - maybe we should be using this (are we?)

**Comment by Tom-Willemsen** [comment]
There is also a `.HVEL` field on the motor record, consider using this instead

**Comment by Tom-Willemsen** [comment]
+1 again from MERLIN as the setting was forgotten

**Comment by JamesKingWork** [comment]
+1 again from MERLIN as the setting was forgotten

**Comment by FreddieAkeroyd** [comment]
Is creep speed only used for homing? Should we set it back to previous value after the home completes?

**Comment by davidkeymer** [comment]
According the the PM600 manual, the _**creep**_ speed is:

> ...the speed that at which moves with a non-zero creep distance will stop. It is also the speed that slow datum search will be moved at (HD command).

**Comment by FreddieAkeroyd** [comment]
I think all moves may have a final number of "creep steps" completed at "creep speed" so we may affect position accuracy if we change it permanently?

**Comment by rerpha** [comment]
PRs:
motor - https://github.com/ISISComputingGroup/EPICS-motor/pull/40
ioc tests - 
release notes -

**Comment by FreddieAkeroyd** [comment]
See comments on motor PR

**Comment by rerpha** [comment]
whoops, realised i forgot to push my emulator changes which i've now somehow lost. Just rewriting now, going to modify the mclennan emulator to care about ID so we can test both models

**Comment by rerpha** [comment]
Have pushed emulator, modified test to test against both versions of mclennan, and modified mclennan code to set creep speed correctly from querying speed and then to set creep speed to VELO when requesting a home.

**Comment by DominicOram** [comment]
@FreddieAkeroyd are you happy to re-review or should someone else?

**Comment by FreddieAkeroyd** [comment]
Am getting emulator/test  errors `'2SV4000', error was: RuntimeError: None of the device's commands matched.` it seems the emulator does not understand the SV command?

**Comment by rerpha** [comment]
i dont think its been implemented in the emulator - does it need to be? I thought we only cared about `SC` - are tests failing?

**Comment by rerpha** [comment]
hmm.. i thought they were passing before, will try and get them running on my machine

**Comment by FreddieAkeroyd** [comment]
The emulator is also not understanding MR that is move relative? It may only understand jog?

**Comment by rerpha** [comment]
Do i need to go and implement those in it then? how on earth was the test passing before?

**Comment by FreddieAkeroyd** [comment]
These may not be relevant, i am just surprised how little the emulator emulates! Does the test pass for you?

**Comment by rerpha** [comment]
It doesn#t, it fails, though im sure i managed to get it to pass

**Comment by FreddieAkeroyd** [comment]
the test that fails is a jog, so MR may not be relevant, but it does mean if you ran up the motors in a simulated ibex you would not be able to move them just jog them. The emulator is certainly being requested to move, but it may have been before when the test passed.

**Comment by FreddieAkeroyd** [comment]
Maybe there is a big and creep speed gets set to zero, and then as it doesnlt know how to set a velocity it never gets set back?

**Comment by FreddieAkeroyd** [comment]
I'll check the log and my logic

**Comment by FreddieAkeroyd** [comment]
no, picking up creep speed of 700 ok. I note emulator has `self.velocity = 0` and if it doesn;t understand a set velocity command i am not sure how it worked...

**Comment by rerpha** [comment]
hmmm. I tried with motor on master and that still failed so think maybe something funny is going on with the emulator

**Comment by FreddieAkeroyd** [comment]
ah, jog sets velocity, but if you tried to move before doing a job maybe you get an issue. Anyway, i think there is some emulator issue and adding `SV` at least would be a good idea

**Comment by rerpha** [comment]
I think the issue is that we are using pm304 as the default model not pm600?

**Comment by rerpha** [comment]
setting this to False makes the tests pass , but i think i may have just uncovered a bug?

https://github.com/ISISComputingGroup/EPICS-DeviceEmulator/pull/120/files#diff-e996946a79e9fc2d620d88593e890cc5c65f92c92a6f8d5d8005c13632ea5a26R20

**Comment by rerpha** [comment]
Have commited that change to https://github.com/ISISComputingGroup/EPICS-DeviceEmulator/pull/120

**Comment by FreddieAkeroyd** [comment]
they pass if is_pm304 is False but still fail if is_pm304 is True - should 304 pass too?

**Comment by rerpha** [comment]
the creep speed tests should pass yes, others im not sure about. maybe I should add that variable into the test as it's not actually being used otherwise?

**Comment by rerpha** [comment]
Actually even if I change that var I'm not sure the motor record will pick it up as it'll be past runtime - i guess it only loads the model number once as it will not change during runtime

**Comment by FreddieAkeroyd** [comment]
I guess you'd either need to pass the variable to the emulator when you launch it, or set it via the backdoor before the ioc is started. either way, you 'd need a different ioc as the motor code reads the ident string at ioc startup and i don't think checks it after that

**Comment by FreddieAkeroyd** [comment]
A bit like danfysik i guess, you want to run the same tests but with the ioc set up in two different ways

**Comment by rerpha** [comment]
As discussed we need to check if anywhere is using the model PM3xx before testing against it in the emulator, which will be done in https://github.com/ISISComputingGroup/IBEX/issues/6792

As far as this ticket is concerned we will leave the device ID variable in the emulator but set it to the PM600 by default. We should also test this ticket's functionality on some real hardware.

**Comment by rerpha** [comment]
OK, have now tested with hardware and it turns out the maximum creep speed you can set is 800. This definitely does speed up homing on the office mclennan by a large amount, the previous value was 50. I have modified the code and added a test so that if HVEL is set to less than 800 the creep speed is set equal to it, and if above 800 the maximum creep speed is applied. Have updated https://github.com/ISISComputingGroup/ibex_developers_manual/wiki/McLennan-motors#homes-are-very-slow with this information.

**Comment by rerpha** [comment]
Going to be a bit cheeky here and merge it rather than wait for another review as the code isn't that different

**Comment by DominicOram** [comment]
I think we've added a whole additional bit of logic, it would be good to have a once over of it

**Comment by rerpha** [comment]
ok, i'll leave it
